walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,105 @@
 
 package com.google.sps;
 
+import java.lang.Math;
+import java.util.HashSet;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.*;
+import com.google.sps.Event;
+import com.google.sps.Events;
+import com.google.sps.MeetingRequest;
+import com.google.sps.TimeRange;
 import java.util.Collection;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    
+    Collection<String> attendees = request.getAttendees();
+    ArrayList<TimeRange> invalidTimes = new ArrayList<>();
+    int fullDay = TimeRange.WHOLE_DAY.duration();
+
+    if(attendees.size() == 0) attendees = request.getOptionalAttendees();   
+    if (request.getDuration() > fullDay) return Arrays.asList();
+
+
+    invalidTimes = generateInvalidTimes(events,attendees);
+    invalidTimes = resolveOverlaps(invalidTimes);
+
+    ArrayList<TimeRange> validTimes = generateValidTimes(invalidTimes, request.getDuration());
+    ArrayList<TimeRange> invalidOptionalTimes = generateInvalidTimes(events, request.getOptionalAttendees());
+
+    
+    for(TimeRange invalid: invalidOptionalTimes) {
+      for(TimeRange valid : validTimes) {
+        if(invalid.equals(valid)) validTimes.remove(invalid);
+      }
+    }
+    TimeRange[] returnValidTimes = new TimeRange[validTimes.size()];
```

Comment:
Did you create an intermediate array in order to create an immutable data structure to return? In this case, `Collections.unmodifiableCollection()` ([javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#unmodifiableCollection-java.util.Collection-)) could help.

Reply:
I tried implementing this, but the tests seems to compare explicitly for Arraylists?
Here's an error I got:
 everyAttendeeIsConsidered(com.google.sps.FindMeetingQueryTest): 
expected: java.util.Arrays$ArrayList<[Range: [0, 480), Range: [510, 540), Range: [570, 1440)]> 
but was: java.util.Collections$UnmodifiableCollection<[Range: [0, 480), Range: [510, 540), Range: [570, 1440)]

So, I'm just going to return the original validTimes list instead